httpcore5/src/main/java/org/apache/hc/core5/net/Host.java

```diff
@@ -71,7 +73,10 @@ static Host parse(final CharSequence s, final Tokenizer.Cursor cursor) throws UR
             if (!InetAddressUtils.isIPv6Address(hostName)) {
                 throw URISupport.createException(s, cursor, "Expected an IPv6 address");
             }
-        } else {
+        }  else if (!StandardCharsets.US_ASCII.newEncoder().canEncode(s.toString())){
```

Comment:
@hanasakari I suspect this operation can be relative expensive. Could you please add a method to `TextUtils` that performs the same test without extra object allocation? 
Please also note that it is likely to be more efficient to execute the test and IDN transformation _after_ `tokenizer#parseContent` parsing operation.